src/test/scala/com/wavesplatform/history/BlockchainUpdaterKeyAndMicroBlockConflictTest.scala

```diff
@@ -0,0 +1,215 @@
+
+          d.blockchainUpdater.processBlock(keyBlock1) shouldBe 'right
+        }
+    }
+
+    forAll(Preconditions.leaseAndLeaseCancel()) {
+      case (genesisBlock, leaseBlock, keyBlock, microBlocks, transferBlock, secondAccount) =>
+        withDomain(MicroblocksActivatedAt0WavesSettings) { d =>
+          Seq(genesisBlock, leaseBlock, keyBlock).foreach(d.blockchainUpdater.processBlock(_) shouldBe 'right)
+          assert(d.blockchainUpdater.effectiveBalance(secondAccount, 0) > 0)
+
+          microBlocks.foreach(d.blockchainUpdater.processMicroBlock(_) shouldBe 'right)
+          assert(d.blockchainUpdater.effectiveBalance(secondAccount, 0, leaseBlock.uniqueId) > 0)
+
+          assert(d.blockchainUpdater.processBlock(transferBlock).toString.contains("negative effective balance"))
+        }
+    }
+  }
+
+  private[this] object TX {
+    val FeeAmount = 400000
+
+    def validTransferGen(from: PrivateKeyAccount, to: AddressOrAlias, amount: Long): Gen[Transaction] =
+      for {
+        timestamp <- timestampGen
+      } yield TransferTransactionV1.selfSigned(None, from, to, amount, timestamp, None, FeeAmount, Array.empty).explicitGet()
+
+    def validLeaseGen(from: PrivateKeyAccount, to: AddressOrAlias, amount: Long): Gen[LeaseTransactionV1] =
+      for {
+        timestamp <- timestampGen
+      } yield LeaseTransactionV1.selfSigned(from, amount, FeeAmount, timestamp, to).explicitGet()
+
+    def validLeaseCancelGen(from: PrivateKeyAccount, leaseId: ByteStr): Gen[LeaseCancelTransactionV1] =
+      for {
+        timestamp <- timestampGen
+      } yield LeaseCancelTransactionV1.selfSigned(from, leaseId, FeeAmount, timestamp).explicitGet()
+  }
+
+  private[this] object Preconditions {
+    import TX._
+
+    def conflictingTransfers(): Gen[(Block, Block, Seq[MicroBlock], Block)] = {
+      for {
+        richAccount   <- accountGen
+        secondAccount <- accountGen
+
+        tsAmount = FeeAmount * 10
+        transfer1 <- validTransferGen(richAccount, secondAccount, tsAmount)
+        transfer2 <- validTransferGen(secondAccount, richAccount, tsAmount - FeeAmount)
+        transfer3 <- validTransferGen(secondAccount, richAccount, tsAmount - FeeAmount)
+      } yield {
+        val genesisBlock = unsafeBlock(
+          reference = randomSig,
+          txs = Seq(GenesisTransaction.create(richAccount, tsAmount + FeeAmount, 0).explicitGet()),
+          signer = TestBlock.defaultSigner,
+          version = 3,
+          timestamp = 0
+        )
+
+        val (keyBlock, microBlocks) = unsafeChainBaseAndMicro(
+          totalRefTo = genesisBlock.signerData.signature,
+          base = Seq(transfer1),
+          micros = Seq(Seq(transfer2)),
+          signer = richAccount,
+          version = 3,
+          timestamp = System.currentTimeMillis()
+        )
+
+        val (keyBlock1, _) = unsafeChainBaseAndMicro(
+          totalRefTo = keyBlock.signerData.signature,
```

Comment:
Could you add a scenario where `micros == Seq(Seq(), Seq(transfer))`, and `keyBlock1` refers to `micros(0)` rather than `keyBlock`?

Reply:
Ok